### ISSUE TITLE
Support NotificationExtenderService with grouped notifications

### ISSUE THREAD

**Issue by xsorifc28** [OneSignal/OneSignal-Android-SDK#68]
I am making local modifications & changes to NotificationCompat.Builder (like circular-cropping the large icon). This works while debugging, I can view the bitmap and it is cropped. The `OneSignalNotificationExtender` does get triggered. The returned `builder` is does not contain the transformed image. I also tried `builder.setLarIcon(croppedBitmap)`, which also did not work. Setting any other properties (small icon, led color, accent color, etc.) also does not work..the `builder` doesn't seem to get modified. 

The weird part is that this worked yesterday..but does not work today..I don't recall changing any parameters.

Any help is appreciated.

Here is the code:

``` Java
    public class OneSignalNotificationExtender extends NotificationExtenderService {
        private static final String TAG = "OSNotificationExtender";

        @Override
        protected boolean onNotificationProcessing(final OSNotificationPayload notification) {
        OverrideSettings overrideSettings = new OverrideSettings();
        overrideSettings.extender = new NotificationCompat.Extender() {
         @Override
         public NotificationCompat.Builder extend(NotificationCompat.Builder builder) {
             Context context = getBaseContext();

             Log.d(TAG,notification.additionalData.toString());

             builder.mLargeIcon = new CircleTransform().transform(builder.mLargeIcon);
             return builder;
             }
         };

        OSNotificationDisplayedResult result = displayNotification(overrideSettings);
        Log.d("OneSignalExample", "Notification displayed with id: " + result.notificationId);

        return true;
        }
    }
```

**Comment by jkasten2** [comment]
@xsorifc28 Did you start sending notifications with a Android group set? The `NotificationExtenderService` doesn't support summary notifications created by grouping yet.

Thanks.

**Comment by xsorifc28** [comment]
Ahh, yes I did. I expected the `NotificationExtenderService` to work with 1 notification though. Are you saying it will still not work if Androud group is set?

**Comment by jkasten2** [comment]
@xsorifc28 When group is set our SDK takes a different code path to display the notification. It has a number of additional complexities so support for this feature was omitted initially to release it sooner.

I am updating this issue to an enhancement and the title to fit.

Thanks.

**Comment by xsorifc28** [comment]
In the meantime, I guess I can group in the `NotificationExtenderService`.

**Comment by fedorcomander** [comment]
Any ETA on this?

**Comment by jkasten2** [comment]
@xsorifc28 @fedorcomander The first notification shown with a group set will respect the `NotificationExtenderService` customization made by it in commit 2c63bc62d6df090fe63137663dc9cf02706dbb78. 2 or more notifications grouped / stacked will still not however.

Also note this limitation won't be an issue on Android N (7.0) devices as the summary notification is no longer shown as it shows it own summary now based on the child notifications attached.

**Comment by kistitan** [comment]
Hi @jkasten2 

Can we reopen this issue please?
It is a great improvement, but our users has device settings for notification ringtone, and they will still hear the incorrect ringtone for the second, third etc... grouped notifications.

Note.... it will take a loot of time until every user will have android N.

As a workaround you can at least get some of the key fields from the extender (sound,color,vibration is at least crucial for us) and add it to the summary builder.
It is quite an easy fix, but would help a lot.
Thanks

**Comment by kistitan** [comment]
To be honest the fix also does not work for me. Extender is still ignored if group is not null.
Also for the first message in the group.

**Comment by kistitan** [comment]
Some suggestion in general: can you be at least be less strict with making all methods final private or package local. It could be an easy override on our side, but because of that I have to make many workarounds to make it work (own build,reflection,same package etc...)

**Comment by kistitan** [comment]
h

**Comment by jkasten2** [comment]
@kistitan I'll try to reproduce the issue with the first notification not respecting the overrides when a group is used. I'll look into the sound issue as well.

For supporting overriding the 2nd, 3rd notification, etc I believe we could read `mContentText` and `mContentTitle` from `NotificationCompat.Build`. Using these 2 values to update the local SQLite records with this so the summary will content your overridden in the summary text.

**Comment by jkasten2** [comment]
@kistitan I wasn't able to reproduce an issue for the first notification displayed part of a group set when overriding the message body and title. However I did reproduce an issue setting a sound from the extender on per-Android 4.3 devices and both played on 4.3 to 6.0.

I just released a 3.5.7 update that fixes the limitation with 2nd, 3rd, etc notifications not respecting the body, title, and sound overrides done in the extender. Note for overriding the sound you will sill need to make sure you disable the `DEFAULT_SOUND` flag otherwise it will override the custom one.
```Java
 builder.mNotification.flags &= ~Notification.DEFAULT_SOUND;
 builder.setDefaults(builder.mNotification.flags);
 return builder.setSound(Uri.parse("content://media/internal/audio/media/32"));
```

Can you try this update to see if you still have an issue? Please provide your code if you still see any issues noting the Android versions you have tested with.

We limit what is marked `public` in our SDK for encapsulation. Open counts, badge counts, which notifications should be restored, and open / received events all depend on the state of notifications. Changing open intent and delete intents for example will case issues with these dependent states. However is is possible we have restricted to much, let us know which private methods you needed to call and I can look into possibly opening this up in an update.

**Comment by txusms** [comment]
Hi. I have same issue with SDK 3.10.7 and Android 6.0.1 and below. Only the first grouped notification plays the custom sound, or not if the user has muted notifications. 2nd, 3rd, etc notifications play default sound. What are I doing wrong?

```kotlin
overrideSettings.extender = object : NotificationCompat.Extender {
   override fun extend(builder: NotificationCompat.Builder): NotificationCompat.Builder {
       return configNotification(builder)
   }
}
...
private fun configNotification(builder: NotificationCompat.Builder): NotificationCompat.Builder {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) return builder
        if (!mRepository.areNotificationsMute()) {
            builder.setDefaults(0)
            builder.setSound(mRepository.notificationsSound())
        } else {
            builder.setDefaults(0)
            builder.setSound(null)
        }
        return builder
    }
```
